### ISSUE TITLE
Tomcat security and API

### ISSUE THREAD

**Issue by unhipzippo** [oracle/opengrok#2872]
(Testing on OpenGrok 1.2.24, AdoptOpenJDK 11.0.3, Tomcat 9.0.21 on Oracle Linux 7.6)

If OpenGrok is running inside a Tomcat instance which uses Tomcat security to protect the entire OpenGrok webapp (rather than using the OpenGrok-specific authorization framework), the Tomcat security defaults to extending to the OpenGrok API as well. Since the OpenGrok indexer's call to the configuration API endpoint doesn't expect/include authorization, this results in indexer completion never informing the webapp of the new configuration.

Somewhat related issues:

https://github.com/oracle/opengrok/issues/2352
https://github.com/oracle/opengrok/issues/2635

In order to permit the indexer to update the webapp with new configuration, https://github.com/oracle/opengrok/wiki/Authorization recommends exempting API calls from security entirely:

```
   <security-constraint>
       <web-resource-collection>                                               
           <web-resource-name>API endpoints are checked separately by the web app</web-resource-name>
           <url-pattern>/api/*</url-pattern>                                   
       </web-resource-collection>                                              
   </security-constraint>
```

However, since the API **search** endpoint can return data which may be considered privileged (e.g. lines of source code), permitting completely public access to the **search** / **suggest** endpoints might not be ideal. (@tulinkry alludes to this in https://github.com/oracle/opengrok/issues/2635#issuecomment-474227453)

Is there a recommended alternative configuration if we want to apply security to API calls?

**Comment by unhipzippo** [comment]
One possibility we're experimenting with is some tweaks to both Tomcat and to our preexisting front-end proxy (nginx):

- Proxy (nginx) on port :80, OpenGrok listening on localhost only, e.g. localhost:8080.
  e.g. Tomcat server.xml:

```
    <Connector address="127.0.0.1" port="8080" protocol="HTTP/1.1"
               connectionTimeout="20000"
               maxHttpHeaderSize="16384"
               redirectPort="8443" />
```

- Add front-end proxy rules to explicitly block all paths under /<webapp_context/api/v1/ EXCEPT for search, suggest and suggest/config endpoints.
  e.g. nginx .../conf/vhosts/opengrok.conf:

```
upstream opengrok-main-http-backend {
    server 127.0.0.1:8080;
    keepalive 10;
    keepalive_timeout 15s;
}

server {
    server_name opengrok.company.com;

    listen *:80;

    location /source/ {

        location ~ ^/source/api/v1/(?!(search|suggest(/config)?)$) {
            allow 127.0.0.1;
            deny all;

            proxy_pass http://opengrok-main-http-backend;
        }

        proxy_pass http://opengrok-main-http-backend;
    }
}
```

- With this proxy change in place, tweak Tomcat security constraint on /source/ webapp to ONLY exempt the API config endpoint from auth enforcement, not all API calls:

```
    <security-constraint>
        <!--
             Explicitly permit /api/v1/configuration without applying our
             below password restriction on it, since the localhost indexer
             needs to call the API without authentication to pass post-index
             config to the webapp.
        -->
        <web-resource-collection>
            <web-resource-name>Auth-exempt area (API config)</web-resource-name>
            <url-pattern>/api/v1/configuration</url-pattern>
        </web-resource-collection>
    </security-constraint>
    <security-constraint>
        <web-resource-collection>
            <web-resource-name>Auth-required area</web-resource-name>
            <url-pattern>/*</url-pattern>
        </web-resource-collection>
        <auth-constraint>
            <role-name>${ldap.group}</role-name>
        </auth-constraint>
    </security-constraint>
    <login-config>
        <auth-method>BASIC</auth-method>
        <realm-name>Username and Password</realm-name>
    </login-config>
    <security-role>
        <role-name>${ldap.group}</role-name>
    </security-role>
```

This should permit the indexer to issue config endpoint calls without requiring authentication (from localhost only), but all other API calls (external or from localhost) should be forced to authenticate via LDAP (and be members of ${ldap.group}, passed in at Tomcat runtime via -Dldap.group)

I haven't yet migrated our instance config to the OpenGrok-specific authorization framework introduced post 1.0, so I don't know offhand if it applies authorization to API calls, or if it similarly permits them without requiring authorization.

**Comment by tulinkry** [comment]
> This should permit the indexer to issue config endpoint calls without requiring authentication (from localhost only), but all other API calls (external or from localhost) should be forced to authenticate via LDAP (and be members of ${ldap.group}, passed in at Tomcat runtime via -Dldap.group)

This way you can use only full index run, not partial index nor [repository synchronization scripts](https://github.com/oracle/opengrok/wiki/Repository-synchronization). But that is not your use case.

> I haven't yet migrated our instance config to the OpenGrok-specific authorization framework introduced post 1.0, so I don't know offhand if it applies authorization to API calls, or if it similarly permits them without requiring authorization.

OpenGrok specific authorization wouldn't solve this for you. It aims to solve per project or per group authorization. The json search request is under this specific authorization when it's in place.

**Comment by vladak** [comment]
> However, since the API search endpoint can return data which may be considered privileged (e.g. lines of source code), permitting completely public access to the search / suggest endpoints might not be ideal. (@tulinkry alludes to this in #2635 (comment))

TLDR: It cannot. The `/search` endpoint is properly authorized.

Long answer: The built in authorization is a bit of a complex beast. There are some related filters defined in the `web.xml` file (like `AuthorizationFilter`) however the `/search` API endpoint is authorized differently. `SearchController` calls `engine.search()` which ends up in https://github.com/oracle/opengrok/blob/39bbbcb2f0a2c86e38ffed912f9c05a07e42b492/opengrok-indexer/src/main/java/org/opengrok/indexer/search/SearchEngine.java#L293-L307 and when the above code calls `pHelper.getAllProjects()` it will eventually bubble down to https://github.com/oracle/opengrok/blob/39bbbcb2f0a2c86e38ffed912f9c05a07e42b492/opengrok-indexer/src/main/java/org/opengrok/indexer/web/ProjectHelper.java#L217-L224 (or `cacheGroups()`) and the actual authorization happens inside `filterProjects()` (or `filterGroups()`) called from there.

**Comment by tulinkry** [comment]
But that is the authorization framework, that is correctly working. But the question was different - the user has LDAP setup to authenticate/authorize users to access the opengrok instance. Which isn't working for /api/v1/search because this endpoint is excluded from this LDAP check as adviced on our wiki.

**Comment by vladak** [comment]
I see. In general I'd deter anyone from doing authorization elsewhere than inside OpenGrok. There are just too many surface points to cover to do this correctly, plus it would be reinventing the wheel. So, from the point of view of the web app, authentication needs to be done externally, authorization internally.

**Comment by tulinkry** [comment]
I also think that the author here mixes authentication with authorization, so what is this about is mainly authentication. It is still valid that unauthenticated user can access /api/v1/search and thus expose the forbidden content.

**Comment by vladak** [comment]
Well, authentication is performed first so that authorization can follow. If authorization was done using the built in framework, it would not permit unauthenticated requests to pass through.

**Comment by vladak** [comment]
That said, the documentation seems to be wrong. I wonder if it is possible to instruct Tomcat to perform authentication only and pass headers to the web application that would then perform authorization. https://stackoverflow.com/questions/7756048/authentication-without-authorization-on-tomcat-7 gives some hope.

**Comment by vladak** [comment]
Adding this to `webapps/source/WEB-INF/web.xml`:
```xml
    <security-constraint>
        <web-resource-collection>
            <web-resource-name>In general everything needs to be authenticated</web-resource-name>
            <url-pattern>/*</url-pattern> <!-- protect the whole application -->
        </web-resource-collection>

        <auth-constraint>
            <role-name>*</role-name>
        </auth-constraint>

        <user-data-constraint>
            <transport-guarantee>NONE</transport-guarantee>
        </user-data-constraint>
    </security-constraint>

    <security-role>
        <role-name>*</role-name>
    </security-role>

    <login-config>
        <auth-method>BASIC</auth-method>
    </login-config>
```
plus having `conf/tomcat-users.xml` to have these contens:
```xml
<?xml version='1.0' encoding='utf-8'?>
<tomcat-users xmlns="http://tomcat.apache.org/xml"
              xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
              xsi:schemaLocation="http://tomcat.apache.org/xml tomcat-users.xsd"
              version="1.0">

  <user username="foo" password="bar" roles="tomcat,manager-script"/>
</tomcat-users>
```
triggers authentication but not authorization in Tomcat.

**Comment by vladak** [comment]
I think we need to introduce proper HTTP Basic Auth decoder for this to be useful. The supplied ` 	HttpBasicAuthorizationPlugin` is for demonstration purposes only.

**Comment by unhipzippo** [comment]
Thanks, @vladak -- I'll try to find the time this week to migrate our previous Tomcat-based authorization config to use the new OpenGrok authorization, using the preexisting LDAP plugin to replace my previous use of Tomcat JNDIRealm / role membership.

**Comment by vladak** [comment]
@tulinkry made be to think of this further - even if the `/search` end point authorizes the requests, one has to explicitly specify the authorizations so there is a bit of a mismatch - while the non API endpoints are not accessible without authentication, the API endpoints are and given the default behaviour of authorization framework the access is allowed by default. I think this can be worked around by using the `FalsePlugin` however maybe a tunable can be introduced to disallow any non-authenticated request (i.e. such that it does not have attribute storing the `User` object).

**Comment by tulinkry** [comment]
I experimented with the tomcat config and it looks like this one should work as expected:

(Note how I explicitly stated the affected endpoints in the authenticated area)

```
    <security-constraint>
        <web-resource-collection>
            <web-resource-name>API endpoints are checked separately by the web app</web-resource-name>
            <url-pattern>/api/*</url-pattern>
        </web-resource-collection>
    </security-constraint>


    <security-constraint>
        <web-resource-collection>
            <web-resource-name>In general everything needs to be authenticated</web-resource-name>
            <url-pattern>/*</url-pattern> <!-- protect the whole application -->
            <url-pattern>/api/v1/search</url-pattern> <!-- protect the search endpoint with basic authentication -->
            <url-pattern>/api/v1/suggest</url-pattern> <!-- protect the suggest endpoint with basic authentication -->
            <url-pattern>/api/v1/suggest/config</url-pattern> <!-- protect the suggest/config endpoint with basic authentication -->
        </web-resource-collection>

        <auth-constraint>
            <role-name>plugins</role-name> <!-- these are the roles from tomcat-users.xml -->
            <role-name>users</role-name> <!-- these are the roles from tomcat-users.xml -->
            <role-name>admins</role-name> <!-- these are the roles from tomcat-users.xml -->
        </auth-constraint>

        <user-data-constraint>
            <!-- transport-guarantee can be CONFIDENTIAL, INTEGRAL, or NONE -->
            <transport-guarantee>NONE</transport-guarantee>
        </user-data-constraint>
    </security-constraint>

    <security-role>
        <role-name>plugins</role-name>
        <role-name>users</role-name>
        <role-name>admins</role-name>
    </security-role>

    <login-config>
        <auth-method>BASIC</auth-method>
    </login-config>
```

**Comment by vladak** [comment]
I think we can do this implicitly - if any authorization stack is defined
in the configuration, check if `User` object exists, otherwise deny the
request.

Dne út 23. 7. 2019 22:16 uživatel Kryštof Tulinger <notifications@github.com>
napsal:

**Comment by tulinkry** [comment]
> I think we can do this implicitly - if any authorization stack is defined in the configuration, check if `User` object exists, otherwise deny the request.

The stacks are user defined. I wouldn't add any additional logic which could cause unexpected behavior when defining custom authorization stacks.

Also, as everything here is custom, what is `User` object? That one is also custom.

**Comment by tulinkry** [comment]
@unhipzippo I think this is answer to your question how to do it with tomcat directly

https://github.com/oracle/opengrok/issues/2872#issuecomment-514366982

**Comment by vladak** [comment]
Added this to the wiki, however I still think we can do better. I have a proof of concept code that introduces `AuthenticationFilter` that ensures http servlet request for these API endpoints has non `null` `getRemoteUser()`. This works as long the stock config is not changed, of course. Also, for non standard authentication methods (such as the one we are using internally that is done outside of Tomcat) a Tomcat Valve would have to be introduced which would set the remote user.

**Comment by vladak** [comment]
Maybe I can use `@PreMatching` like in `LocalhostFilter` and embed the API URLs in the `AuthenticationFilter`. This way it will be immune against `web.xml` tinkering.

**Comment by vladak** [comment]
Because of the complications (need for Tomcat Valve + unclear modifications to other application servers) I saved the work in https://github.com/vladak/OpenGrok/tree/authentication_filter for now.

**Comment by tulinkry** [comment]
>  I have a proof of concept code that introduces AuthenticationFilter that ensures http servlet request for these API endpoints has non null getRemoteUser()

But why? When user has not configured any auth, then even standard pages (xrefs, history, ...) don't have the non null`getRemoteUser`. Why there would be such restriction on using the API?

This non null `getRemoteUser` is **enforced** by using the Tomcat auth with the changes I posted before (to include /api/v1/search, suggest, suggest/config).

**Comment by vladak** [comment]
The trouble is that `web.xml` is overwritten with default contents on redeploy.

**Comment by tulinkry** [comment]
> The trouble is that web.xml is overwritten with default contents on redeploy.

Yes, however that's how you do it with tomcat. In our case people already found the way to unpack the war and add their customizations and pack it back.

**Comment by vladak** [comment]
Suddenly I find this quite error prone. Having the webapp to check that authorization can be actually put to use would be a safety net.

**Comment by vladak** [comment]
Maybe we can actually make the `opengrok-deploy` script to insert some contents into `web.xml` given that it already changes it anyway.

**Comment by unhipzippo** [comment]
Thanks, @tulinkry . Initially I wondered whether your combined recommendation of **/api/v1/suggest/*** might miss that basic **suggest** endpoint (since it doesn't end in a slash, and I was thinking in regex terms).

https://github.com/oracle/opengrok/issues/2872#issuecomment-514515802

However, my testing indicates it **does** correctly force authentication for both **/suggest?foo** and **suggest/config**. (Is this due to some intricacy of the URL path rules in the Java servlet spec?)

**Comment by unhipzippo** [comment]
>> The trouble is that web.xml is overwritten with default contents on redeploy.

> Yes, however that's how you do it with webapp containers. In our case people already found the way to unpack the war and add their customizations and pack it back.

Agreed -- that's what our deployment does -- Any changes we need to make to the webapp content (e.g. web.xml) are made as part of an Ansible deployment playbook, which patches and repackages source.war as needed before deploying it to destination machines.

**Comment by vladak** [comment]
I have split and rewritten the Authorization wiki page to match the newly added plugins so I assume this issue can be closed.

**Comment by vladak** [comment]
If not, feel free to reopen.
